cartographer_grpc/framework/service.cc

```diff
@@ -91,24 +101,56 @@ void Service::HandleNewConnection(Rpc* rpc, bool ok) {
                               rpc->server_completion_queue());
 }
 
-void Service::HandleDone(Rpc* rpc, bool ok) { LOG(FATAL) << "Not implemented"; }
+void Service::HandleRead(Rpc* rpc, bool ok) {
+  if (ok) {
+    rpc->OnRequest();
+    RequestStreamingRead(rpc);
+    return;
+  }
+
+  // Reads completed.
+  rpc->OnReadsDone();
+}
+
+void Service::HandleWrite(Rpc* rpc, bool ok) {
+  if (!ok) {
+    LOG(ERROR) << "Write failed";
+  }
+
+  CleanUpAsNeeded(rpc);
+}
+
+void Service::HandleDone(Rpc* rpc, bool ok) { CleanUpAsNeeded(rpc); }
+
+void Service::CleanUpAsNeeded(Rpc* rpc) {
+  if (!rpc->IsRpcStatePending(Rpc::State::DONE) &&
+      !rpc->IsRpcStatePending(Rpc::State::READ) &&
+      !rpc->IsRpcStatePending(Rpc::State::WRITE)) {
+    active_rpcs_.Remove(rpc);
+  }
+}
 
 void Service::RequestNextMethodInvocation(
     int method_index, Rpc* rpc,
     ::grpc::ServerCompletionQueue* completion_queue) {
   rpc->server_context()->AsyncNotifyWhenDone(
-      rpc->GetRpcState(Rpc::State::DONE));
+      rpc->SetRpcStatePending(Rpc::State::DONE, true));
   switch (rpc->rpc_type()) {
     case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
-      RequestAsyncClientStreaming(method_index, rpc->server_context(),
-                                  rpc->streaming_interface(), completion_queue,
-                                  completion_queue,
-                                  rpc->GetRpcState(Rpc::State::NEW_CONNECTION));
+      RequestAsyncClientStreaming(
```

Comment:
Service needs several of rpc's attributes. Consider alternative designs such as making Rpc a small struct without logic and doing the work in Service.